apps/engine/lib/execution/Cargo.toml

```diff
@@ -21,6 +21,8 @@ rand = "0.8.5"
 rayon = "1.5.3"
 serde = { version = "1.0.138", features = ["derive"] }
 serde_json = "1.0.82"
+swc = "0.261.29"
+swc_common = "0.31.11"
```

Comment:
Is this also possible using `v8`? `swc` is a pretty big dependency with very many (and outdated) dependencies.

Reply:
There was a [related PR](https://github.com/hashintel/hash/pull/567), which added TS support for another location in the code. I haven't looked at transpiling TS to JS code in Rust, yet, but it might worth taking a look 🙂

Reply:
No, v8 does not have TypeScript compatibility. Unfortunately, the options to compile TypeScript to JavaScript are pretty limited and all pretty heavy in this context. swc is used all throughout the TS ecosystem, perhaps most notably by Next.js, and is often touted as the fastest/lightest production compiler. It's also the only compiler I know of that doesn't require installing it externally and shelling out to it.

I'm not a huuuge fan of swc either, but it's kind of the best for what it does right now. Competitors include esbuild (which includes bundling), babel, and the default TypeScript compiler (`tsc`). Eventually, `tsc` needs to be in the stack somewhere to do actual type-checking, but the compiler is written in JS, it's quite slow (seconds where swc is milliseconds), and it isn't easily invoked directly through Rust (unlike swc). Additionally, the next stop here was to add `swc_bundler`, the best bundling solution I can see (can turn two TS files into one; this would allow importing files from other files, including JS) that's accessible by Rust.

Another solution I've considered is transitioning the JS runner to use [`bun`](https://bun.sh/) which can do a lot of stuff "automatically". Downsides include lack of Windows support, it's supported by a single developer, and we'd definitely have to shell out to invoke it (rather than being able to inline it). However, I think there'd be a lot of upsides including automatically included ffi. There's potentially a lot to discuss here.

I'm definitely open to other solutions, though. I just think they're mostly actually heavier and more cumbersome than `swc`. I also don't want to keep investing in swc (through `swc_bundler`) if you don't want to merge it here, so definitely worth figuring out what the optimal final setup would look like. In particular, the alternatives to `swc_bundler` would all require shelling out as well, so I think we could want a totally different approach.

I'd boil it down to the two approaches of

1. Do everything inline/in memory (the way this PR does). If we want this, there aren't really any good alternatives to swc right now, but if they ever were developed (e.g. Rome), we could pretty easily swap it in place of swc.
2. Shell out to another project that uses the file system itself. This would be something like `esbuild` or `tsc`. I view this as the "heavier" option, but one nice aspect is that I think it will be more naturally conducive to bundling as one step (`esbuild` in particular can do both).

Please feel free to take your time with this. I'm in no rush to get this merged in, and I don't think you are either. I'm pretty happy either way, but I would of course eventually like to be able to write simulations using TypeScript, so that's my only real goal.

Reply:
Cheers for the detailed and thoughtful reply.

FWIW, I know Jarred @ Oven raised $7m+ for [`bun`](https://bun.sh/) and friends quite a while back now, so if it is genuinely just supported by him at the moment I think that likely that won't be the case for all that long. But there is certainly more adoption risk w/ it than `swc` in my mind, and I would consider lack of native Windows support to be a blocker given our existing/historical user-base and desire to remain widely accessible.

I'll leave actual follow-up re: next steps and a decision here to Tim, but I can see the argument for `swc` and the benefit this PR provides beyond what we have in place already. Like you say, updates to this version of hEngine are not a high priority for us at present, but this seems like an obvious win if we can bag it in some form, and we're very grateful for the contribution here. We'll do what we can to see it reviewed in a timely fashion.

Reply:
Yep, I'm very excited about bun in the future. 

For my edification, what are some of your top priorities at the moment? Is there a future version of the engine in the works?

Reply:
@rivertam We're presently evaluating the best role for hEngine to play in simulation pipelines that more heavily make use of generative agents. For the avoidance of doubt, we aim to continue work on this version of the engine (as opposed to another), but the future roadmap is currently in flux.

Reply:
Yes, exactly. :) (And sorry, missed this comment somehow!)